test/e2e/framework/service_util.go

```diff
@@ -1192,12 +1196,13 @@ func ValidateEndpointsOrFail(c clientset.Interface, namespace, serviceName strin
 	Failf("Timed out waiting for service %s in namespace %s to expose endpoints %v (%v elapsed)", serviceName, namespace, expectedEndpoints, ServiceStartTimeout)
 }
 
-// StartServeHostnameService creates a replication controller that serves its hostname and a service on top of it.
-func StartServeHostnameService(c clientset.Interface, internalClient internalclientset.Interface, ns, name string, port, replicas int) ([]string, string, error) {
+// StartCustomizedServeHostnameService creates a replication controller that
+// serves its hostname and a service on top of it. Caller of the function can
+// provide a tweak function to modify the configuration of the service before
+// it is created.
+func StartCustomizedServeHostnameService(c clientset.Interface, internalClient internalclientset.Interface, ns, name string, port, replicas int, tweak func(svc *v1.Service)) ([]string, string, error) {
```

Comment:
Can we just pass the Service into the function, rather than using tweak()? I understand it is ok in this context, but I don't want to create a pattern that could potentially become quite confusing.

You can use Clone() to copy the various API objects if you need to start with a base that has some settings and change values:
```
svc := &v1.Service { ... }

copy := svc.Clone()
copy.Spec.Type = ...
```

Reply:
Sure I can do that. Now service will be passed as parameter and be created outside of StartServeHostnameService.
See if I did it right way.